### PR TITLE
Parse alias filters on the coordinating node

### PR DESCRIPTION
**Pull request by s1monw**
Today we don't parse alias filters on the coordinating node, we only forward
the alias patters to executing node and resolve it late. This has several problems
like requests that go through filtered aliases are never cached if they use date math,
since the parsing happens very late in the process even without rewriting. It also used
to be processed on every shard while we can only do it once per index on the coordinating node.
Another nice side effect is that we are never prone to cluster-state updates that change an alias,
all nodes will execute the exact same alias filter since they are process based on the same
cluster state.

### CODE REVIEW

**Comment by s1monw** [comment]
@colings86 can you take a look

File: core/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 1]
```diff
@@ -76,4 +89,65 @@
...
+            if (alias == null) {
+                // This shouldn't happen unless alias disappeared after filteringAliases was called.
+                throw new InvalidAliasNameException(index, aliasNames[0], "Unknown alias name was passed to alias Filter");
+            }
+            return parse(index, alias, contextFactory);
+        } else {
+            // we need to bench here a bit, to see maybe it makes sense to use OrFilter
+            BoolQueryBuilder combined = new BoolQueryBuilder();
+                for (String aliasName : aliasNames) {
+                AliasMetaData alias = aliases.get(aliasName);
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java]: Looks like there is an indentation issue here?

**Review by colings86 — APPROVED** [review]
Left one tiny comment but LGTM

**Comment by colings86** [comment]
Looks like org.elasticsearch.search.SearchServiceTests.testSearchWhileIndexDeleted failed in the PR build. Not sure if this failure is related to this change?

**Comment by s1monw** [comment]
@colings86 I added additional testing can you take another look?

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 2]
```diff
@@ -71,6 +72,19 @@ public TransportSearchAction(Settings settings, ThreadPool threadPool, BigArrays
...
         SearchTransportService.registerRequestHandler(transportService, searchService);
         this.clusterService = clusterService;
+        this.searchService = searchService;
+    }
+
+    private Map<String, AliasFilter> buildPerIndexAliasFilter(SearchRequest request, ClusterState clusterState, String...concreteIndices) {
+        final Map<String, AliasFilter> queryBuilderMap = new HashMap<>();
+        for (String index : concreteIndices) {
+            clusterState.blocks().indexBlockedRaiseException(ClusterBlockLevel.READ, index);
+            AliasFilter queryBuilder = searchService.buildAliasFilter(clusterState, index, request.indices());
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: nit: should this be called `aliasFilter`?

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 3]
```diff
@@ -71,6 +72,19 @@ public TransportSearchAction(Settings settings, ThreadPool threadPool, BigArrays
         this.searchTransportService = new SearchTransportService(settings, transportService);
         SearchTransportService.registerRequestHandler(transportService, searchService);
         this.clusterService = clusterService;
+        this.searchService = searchService;
+    }
+
+    private Map<String, AliasFilter> buildPerIndexAliasFilter(SearchRequest request, ClusterState clusterState, String...concreteIndices) {
+        final Map<String, AliasFilter> queryBuilderMap = new HashMap<>();
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: nit: should this be called `aliasFilterMap`?

**Review by colings86 — APPROVED** [review]
Left 2 nit pick comments but LGTM

**Comment by javanna** [comment]
@s1monw I think that we should add a bw comp layer on top of this change, which otherwise breaks backwards compatibility on the transport layer. Maybe our bw comp tests don't catch it because we have no tests that search, or validate query, or explain against a filtered alias?

**Comment by javanna** [comment]
@javanna you were wrong, this is fully bw compatible :)

**Comment by s1monw** [comment]
> @s1monw I think that we should add a bw comp layer on top of this change, which otherwise breaks backwards compatibility on the transport layer. Maybe our bw comp tests don't catch it because we have no REST tests that search, or validate query, or explain against a filtered alias?

the bwc layer is inside the new class... it's not very obvious from the high level but when you look closer you see that all kinds of BWC tests are present.

**Pull request merged**